### PR TITLE
Lazy merkle trees

### PR DESCRIPTION
**Pull request by kustosz**

### CODE REVIEW

**Review by gswirski — COMMENTED** [review]
1. I'd really love to separate types for the original tree (the one on which we can call `update_with_mutation`) from types that can only be updated through sparse nodes.
2. It needs tests with large trees to prevent regressions and ensure we're not allocating massive amounts of memory by accident.

Future work: provide a routine to consolidate sparse nodes back into a dense tree.

File: src/lazy_merkle_tree.rs [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+        proof.root(value) == self.root()
+    }
+
+    /// Sets the value at the given index to the given value. This is a mutable
+    /// operation, that will modify any dense subtrees in place.
+    ///
+    /// This has potential consequences for the soundness of the whole
+    /// structure: 1. `self` is only valid after calling this if the tree it
+    /// represents is fully dense or fully sparse. If it is a mix, it will
+    /// be left in an invalid state.
```
Comment (iamrecursion) [inline src/lazy_merkle_tree.rs:74]: Can we not do better than this? Is there a way to detect whether it'll get into an invalid state cheaply?

File: src/lazy_merkle_tree.rs [hunk 2]
```diff
@@ -0,0 +1,769 @@
...
+        Turn::Right
+    }
+}
+
+const fn clear_turn_at_depth(index: usize, depth: usize) -> usize {
+    index & !(1 << (depth - 1))
+}
+
+impl<H: Hasher> From<Children<H>> for SparseTree<H> {
+    fn from(children: Children<H>) -> Self {
```
[the remark this answers is not in the document]
Comment (iamrecursion) [reply src/lazy_merkle_tree.rs:345]: Assert, or make it fallible?

File: src/merkle_tree.rs [hunk 3]
```diff
@@ -14,7 +14,7 @@ use std::{
 /// Hash types, values and algorithms for a Merkle tree
 pub trait Hasher {
     /// Type of the leaf and node hashes
-    type Hash: Clone + Eq + Serialize;
+    type Hash: Clone + Eq + Serialize + Debug;
```
[the remark this answers is not in the document]
Comment (iamrecursion) [reply src/merkle_tree.rs:17]: A worthwhile one in my opinion, but yes.

**Review by iamrecursion — APPROVED** [review]
LGTM. Some comments in line, and one question.

I don't know if we do better, but always reading this API I kept getting caught on the naming with regards to "mutation". Yes, you are mutating under the hood, but also you never actually take `&mut self`, which makes it a little less clear. That said, I _really_ don't see a way to do better, but thought I'd raise the concern regardless.

File: src/lazy_merkle_tree.rs [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+        children.into()
+    }
+
+    fn root(&self) -> H::Hash {
+        self.root.clone()
+    }
+
+    fn get_leaf(&self, index: usize) -> H::Hash {
+        self.children.as_ref().map_or_else(
+            || self.root.clone(),
```
[the remark this answers is not in the document]
Comment (kustosz) [reply src/lazy_merkle_tree.rs:432]: WDYM non-existent? That's just the termination condition – there's no children, we're at the leaf, we found the result. There is no error to be handled here.

File: src/lazy_merkle_tree.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kustosz) [reply src/lazy_merkle_tree.rs:345]: assert, this is internal

File: src/lazy_merkle_tree.rs (hunk 1 above)
Comment (kustosz) [reply src/lazy_merkle_tree.rs:74]: Nevermind, I just made this drop self

File: src/lazy_merkle_tree.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gswirski) [reply src/lazy_merkle_tree.rs:491]: Nvm, didn’t notice recursion. Looks good.

**Review by gswirski — APPROVED** [review]
I really like the `Canonical` vs `Derived` change. 👍

**Pull request merged**